### PR TITLE
feat: Support `import db from './my.db' with {type: "sqlite"}`

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

This implements `import` support for SQLite databases.

```js
import db from './my.db' with {type: "bun"};

db.query("SELECT * from users LIMIT 1");
```

This also implements support for embedding SQLite databases with `bun build --compile`.

```js
import db from './my.db' with {type: "bun", embed: "true"};

db.query("SELECT * from users LIMIT 1;");
```

 Also, fixed bundling .node files 

Fixes https://github.com/oven-sh/bun/issues/5489
Fixes https://github.com/oven-sh/bun/issues/3653
Fixes https://github.com/oven-sh/bun/issues/3903

### How did you verify your code works?

Tests

### CODE REVIEW

File: src/bun.js/module_loader.zig [hunk 1]
```diff
@@ -2277,6 +2327,26 @@ pub const ModuleLoader = struct {
             const specifier_utf8 = specifier.toUTF8(bun.default_allocator);
             defer specifier_utf8.deinit();
             if (graph.files.get(specifier_utf8.slice())) |file| {
+                if (file.loader == .sqlite or file.loader == .sqlite_embedded) {
+                    const code =
+                        \\/* Generated code */
+                        \\import {Database} from 'bun:sqlite';
+                        \\import {readFileSync} from 'node:fs';
+                        \\export const db = new Database(readFileSync(import.meta.path));
```
Comment (Electroid) [inline src/bun.js/module_loader.zig:2335]: Is this suppose to be the same for embeded and non-embeded?

**Review by Electroid — APPROVED** [review]
Nice

File: src/bun.js/module_loader.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bun.js/module_loader.zig:2335]: yeah, readFileSync will do the right thing

**Comment by codehz** [comment]
maybe also add support for open flags?
like use SQLITE_OPEN_URI (so we can use it as named shared in-memory database (file:xxxx?mode=memory&cache=shared), which can be shared between macro and build script)
or just mapping more argument to attribute
```
import db from "test" with { type: "sqlite", mode: "memory", cache: "shared" }
```
mode=memory = [SQLITE_OPEN_MEMORY](https://www.sqlite.org/c3ref/c_open_autoproxy.html)
cache=shared = [SQLITE_OPEN_SHAREDCACHE](https://www.sqlite.org/c3ref/c_open_autoproxy.html)
etc

**Comment by Jarred-Sumner** [comment]
> maybe also add support for open flags? like use SQLITE_OPEN_URI (so we can use it as named shared in-memory database (file:xxxx?mode=memory&cache=shared), which can be shared between macro and build script) or just mapping more argument to attribute
> 
> ```
> import db from "test" with { type: "sqlite", mode: "memory", cache: "shared" }
> ```
> 
> mode=memory = [SQLITE_OPEN_MEMORY](https://www.sqlite.org/c3ref/c_open_autoproxy.html) cache=shared = [SQLITE_OPEN_SHAREDCACHE](https://www.sqlite.org/c3ref/c_open_autoproxy.html) etc

Maybe in the future, but not right now

File: src/bundler/bundle_v2.zig [hunk 2]
```diff
@@ -2501,6 +2526,116 @@ pub const ParseTask = struct {
                 }, Logger.Loc{ .start = 0 });
                 return JSAst.init((try js_parser.newLazyExportAST(allocator, bundler.options.define, opts, log, root, &source, "")).?);
             },
+
+            .sqlite_embedded, .sqlite => {
+                if (!bundler.options.target.isBun()) {
+                    log.addError(
+                        null,
+                        Logger.Loc.Empty,
```
Comment (paperclover) [inline src/bundler/bundle_v2.zig:2534]: this not having a location is lame.

image

someone will eventually run into this error, where they want to find and remove the import.

**Review by paperclover — CHANGES_REQUESTED** [review]
awesome. not yet ready to ship imo.

- `bun build` without --compile will emit the same import statement with the `with` clause. this is fine but i was expecting the wrapper to be bundled in. (no action needed on this point)
- `embed: "true"` does not work outside of --compile. this should either be an error, or the generated code should just be a top level import statement like it is without embed
- `To use the "sqlite" loader, set target to "bun"` this message has no error location and will confuse people. i hope it is possible to fix this, but if not i will open an issue about it.

**Comment by Jarred-Sumner** [comment]
> embed: "true" does not work outside of --compile. this should either be an error, or the generated code should just be a top level import statement like it is without embed

It does work outside of `--compile`. It is copied into the executable and loaded at runtime. It becomes a no-op at runtime. There is a test for this.

> bun build without --compile will emit the same import statement with the with clause. this is fine but i was expecting the wrapper to be bundled in. (no action needed on this point)

We rely on loading it at runtime for this

File: src/bundler/bundle_v2.zig (hunk 2 above)
Comment (Jarred-Sumner) [reply src/bundler/bundle_v2.zig:2534]: could add the error earlier i think

**Pull request merged**